File: 2204-find-subsequence-of-length-k-with-the-largest-sum/find-subsequence-of-length-k-with-the-largest-sum.cpp
class Solution {
public:
    vector<int> maxSubsequence(vector<int>& nums, int k) {
        int sz = nums.size(), j=0;
        vector<pair<int, int>> tmp;
        vector<int> ans(k, 0);

        for(int i=0; i<sz; i++){
            tmp.push_back({nums[i], i});
        }

        sort(tmp.begin(), tmp.end(), [](pair<int, int>& a, pair<int, int>& b){
            return a.first > b.first;
        });

        sort(tmp.begin(), tmp.begin()+k, [](pair<int, int>& a, pair<int, int>& b){
            return a.second < b.second;
        });

        for(int i=0; i<k; i++){
            ans[i] = tmp[i].first;
        }

        return ans;
    }
};

File: 2204-find-subsequence-of-length-k-with-the-largest-sum/find-subsequence-of-length-k-with-the-largest-sum.java
class Solution {
    public int[] maxSubsequence(int[] nums, int k) {
        int sz = nums.length, j = 0;
        int ans[] = new int[k];
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> {
            return nums[a] - nums[b];
        });

        for(int i=0; i<sz; i++){
            if(pq.size() < k) pq.add(i);
            else if(nums[pq.peek()] < nums[i]){
                pq.poll();
                pq.add(i);
            }
        }

        for(int i=0; i<sz; i++){
            if(pq.contains(i)){
                ans[j++] = nums[i];
                pq.remove(i);
            }
        }

        return ans;
    }
}